### PR TITLE
Refactor AnswerDao and related tests

### PR DESCRIPTION
**Pull request by yufengwng**
## Context and the big picture

Finally. There is now only one `AnswerDao` and it's all JDBI code.

Summary:
* Centralized to one `AnswerDao` and `AnswerSql`
* Removed the other tiny `Jdbi*` answer files
* Updated all the tests that uses `AnswerDao` in one form or another
 
 ## What type of change is this?
  
  - [x] Bugfix (non-breaking change which fixes an issue)
  - [ ] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [ ] Experimental proof-of-concept
  - [ ] Infosec/Compliance remediation
  - [x] Refactoring
  
 ## Risk Assessment
  - [x] These changes are **HIGH RISK**.
    - [x] There is an elevated risk that may impact a large number of features
    - [ ] There is an elevated risk to security and privacy
    - [ ] These changes introduce new 3rd party dependencies, alter the network perimeter, or contain
    major upgrades to persistent storage or SaaS dependencies such as auth0, sendgrid, or easypost    
  - [ ] These changes are low risk, do not impact security/privacy, do not impact pepper shared
  components/modules, and have no major upgrades to 3rd party services or libraries
 
 ### FUD Score 
 Overall, how are you feeling about these changes?
  - [x] :relaxed: All good, business as usual!
  - [ ] :sweat_smile: There might be some issues here
  - [ ] :scream:I'm sweaty and nervous
 
 ## How to we demo these changes?
 How does one observe these changes in a deployed system?  Note that _user visible_ encompasses
 many personas--not just patients and study staff, but ops your fellow devs, compliance, etc.
 
 - [x] They are user-visible in dev as a regular user journey and require no additional instructions.
 - [ ] Getting dev into a state where this is user-visible requires some tech fiddling.  I have
 documented these steps in the related ticket.
 - [ ] Requires other features before it's human visible.  I have documented the blocking issues
 in jira.
 - [ ] I have no idea how to demo this.  Please help me!
 
 ## UI/UX
  - [x] There ain't no UI/UX in these changes
  - [ ] My changes have passed muster with Erin and Jen via an over-the-shoulder review, screenshots, etc.
  - [ ] Erin and Jen have not had an opportunity to provide feedback yet
 
###  Backend
   - [ ] I have considered error handling and notification to slack alert rooms via `LOG.error()`.
     #### Route Changes
     - [ ] My changes involve a new route changes to an existing route, so the who/what/when are logged using the usual logback mechanism
at the _start_ of the route, prior to any validation, as well as at the end of the route when work completes.
     ### Housekeeping Changes
     - [ ] My changes involve updates to housekeeping, and I am logging the who/what/when via logback.
  ### Client-side
   - [ ] My changes do not involve backend route change, but I have useful logging statements nonetheless.

## Analytics
 - [ ] I have discussed the analytics needs at both a platform and study level with Jen and instrumented code
 accordingly.
 - [x] There are no analytics requirements for these changes

## Security and Privacy
Ensuring proper security and privacy for our users is essential.

  ###  Backend
   - [ ] These changes alter fundamental auth filter logic
   - [ ] These changes introduce a new backend API route that is behind an auth filter
   - [ ] These changes introduce a new backend API that _does not require auth_
   - [ ] These changes alter auth0 internals (rules, hosted login, configuration, etc.)
   - [ ] These changes expose new information to elastic and I have verified that PHI is not being exposed to the public
    
  ### Browser
  - [ ] These changes alter what we store in browser storage
  
  ### Mobile
  - [ ] TBD 
 
 ## Performance and Stability
  - [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, and performance bottlenecks and written a brief summary in this PR
  - [ ] I'm unsure about stability, performance, fault tolerance, and graceful degradation.  Please help!
  
## Testing
 - [x] I have written automated positive tests
 - [x] I have written automated negative tests
 - [ ] I have written zero automated tests but have poked around locally to verify proper functionality
 - [ ] The jira ticket has acceptance criteria and Kiara has in the information she needs to test my changes
 
## Release
 - [x] These changes require no special release procedures--just code!
 - [ ] Releasing these changes requires special handling
   - [ ] I have documented the special procedures in the release plan document
   - [ ] The special procedures require minimal downtime
   - [ ] The special procedures require unavoidable, extended downtime for:
     - [ ] Participant UX
     - [ ] Study staff

### CODE REVIEW

File: ddp-automation/src/test/java/org/broadinstitute/ddp/tests/DatabaseUtility.java [hunk 1]
```diff
@@ -105,8 +105,8 @@ private static void listResultSet(List<Long> resultSet) {
     public static List<Long> getAnswerIds() {
         List<Long> ids = TransactionWrapper.withTxn(handle -> {
             logger.info("Using user id: {}", getUserId());
-            JdbiAnswer answerDao = handle.attach(JdbiAnswer.class);
-            return answerDao.getAnswerIds(getUserId());
+            AnswerSql answerSql = handle.attach(AnswerSql.class);
+            return List.copyOf(answerSql.findAllAnswerIdsByOperatorUserId(getUserId()));
```
Comment (yufengwng) [inline ddp-automation/src/test/java/org/broadinstitute/ddp/tests/DatabaseUtility.java:109]: @Rinchan20 this should provide the same functionality as what we used here before.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/AnswerDao.java [hunk 2]
```diff
@@ -49,212 +47,38 @@
...
-                    return answer;
-                } else {
-                    throw new DaoException("Unable to find CompositeAnswer with id" + answerId);
-                }
-            default:
-                throw new DaoException("Unhandled question type " + questionType);
-        }
+    // Convenience method for creating answers using guids. Prefer the other method that takes ids.
+    default Answer createAnswer(String operatorGuid, String instanceGuid, Answer answer) {
+        long operatorId = getHandle().attach(JdbiUser.class).getUserIdByGuid(operatorGuid);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/AnswerDao.java:65]: Don't know what to do about this id vs guid business. Ideally we won't have to do these extra queries just to convert between id/guid, but not sure what else to do here.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/AnswerDao.java [hunk 3]
```diff
@@ -325,6 +320,16 @@ public void accumulate(Map<Long, Answer> container, RowView view) {
             reduce(container, view);
         }
 
+        @Override
+        public Stream<Answer> stream(Map<Long, Answer> container) {
+            if (!childAnswers.isEmpty()) {
+                // There are child answers that has not been consumed by a parent answer,
+                // very likely we're querying for the child answers themselves so return them.
+                container.putAll(childAnswers);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/AnswerDao.java:328]: Found an issue where we were not able to fetch the individual child answers separately from the parent. This should fix it.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/AnswerDao.java [hunk 4]
```diff
@@ -300,23 +128,190 @@ private void createAnswerPicklistValue(long instanceId, long answerId, PicklistA
...
+        deleteAnswers(getAnswerSql().findAllAnswerIdsByInstanceIds(instanceIds));
+    }
+
+    //
+    // queries
+    //
+
     @UseStringTemplateSqlLocator
-    @SqlQuery("queryAnswerById")
+    @SqlQuery("findAnswerById")
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/AnswerDao.java:280]: Note that I am using the same name for the sql query as the method name. Seems more consistent and less silly this way.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiCompositeAnswer.java [hunk 5]
```diff
@@ -89,7 +90,8 @@
                                                 rs.getString("child_answer_guid"),
                                                 rs.getLong("child_question_id"),
                                                 rs.getString("child_question_stable_id"),
-                                                QuestionType.valueOf(rs.getString("child_question_type_code"))));
+                                                QuestionType.valueOf(rs.getString("child_question_type_code")),
+                                                rs.getLong("parent_activity_instance_id")));
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiCompositeAnswer.java:94]: It should be safe to assume that child answers have the same `activity_instance_id` as the parent. CC @MocanaAtBroad

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistAnswerDao.java [hunk 6]
```diff
@@ -94,13 +80,9 @@ default void assignOptionsToAnswerId(long answerId, List<SelectedPicklistOption>
...
         }
 
-        long[] ids = getJdbiPicklistOptionAnswer().bulkInsert(answerId, selectedIds, detailTexts);
+        long[] ids = getAnswerSql().bulkInsertPicklistSelected(answerId, selectedIds, detailTexts);
         if (ids.length != selected.size()) {
             throw new DaoException("Not all selected picklist options were assigned to answer " + answerId);
         }
     }
-
-    default int unassignOptionsFromAnswerId(long answerId) {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistAnswerDao.java:103]: This file is a bit complicated, so I just moved this method to `AnswerSql` and kept the other method here.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiCompositeAnswer.java [hunk 7]
```diff
@@ -69,6 +69,7 @@
                                 parentAnswer.setId(rs.getLong("parent_answer_id"));
                                 parentAnswer.setGuid(rs.getString("parent_answer_guid"));
                                 parentAnswer.setQuestionStableId(rs.getString("parent_question_stable_id"));
+                                parentAnswer.setActivityInstanceId(rs.getLong("parent_activity_instance_id"));
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiCompositeAnswer.java:72]: I'm a bit afraid to touch the composite answer code, so I left this `JdbiCompositeAnswer` file.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/PatchFormAnswersRoute.java [hunk 8]
```diff
@@ -199,39 +203,40 @@ public Object handle(Request request, Response response) {
                         }
                         failedRules.addAll(failures);
                     } else {
-                        // todo: proper usage of flag when requirements are better understood
-                        boolean forceAdd = false;
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/PatchFormAnswersRoute.java:203]: No one has mentioned anything about this in past few years. I'm removing it.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/StudyDataLoader.java [hunk 9]
```diff
@@ -727,18 +725,18 @@ public String answerCompositeQuestion(Handle handle,
...
-        String guid = answerDao.createAnswer(handle, answer, participantGuid, instanceGuid);
+        Answer parentAnswer = answerDao.createAnswer(participantGuid, instanceGuid,
+                new CompositeAnswer(null, pepperQuestionStableId, null));
         List<Long> childrenAnswerIds = new ArrayList<>();
 
+        var jdbiCompositeAnswer = handle.attach(JdbiCompositeAnswer.class);
         if (CollectionUtils.isNotEmpty(nestedGuids)) {
             for (String childGuid : nestedGuids) {
-                childrenAnswerIds.add(answerDao.getAnswerIdByGuids(handle, instanceGuid, childGuid));
+                childrenAnswerIds.add(answerDao.getAnswerSql().findDtoByGuid(childGuid).get().getId());
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/StudyDataLoader.java:735]: I think `DataLoader` and `StudyDataLoader` are like the only places that uses the answer guids extensively. Might be nice to switch to passing around answer ids or `Answer` objects to get rid of this extra query. CC @ssettipalli

**Comment by yufengwng** [comment]
please run tests

**Comment by yufengwng** [comment]
please run tests

**Review by raiks2 — APPROVED** [review]
Looks good to me

**Pull request merged**